app/payment-request-mapping/index.js

```diff
@@ -0,0 +1,35 @@
+const db = require('../data')
+const paymentRequestSchema = require('./payment-request-schema')
+
+async function savePaymentRequest (paymentRequest) {
+  const validationResult = paymentRequestSchema.validate(paymentRequest)
+
+  if (validationResult.error) {
+    throw new Error(`Payment request is invalid. ${validationResult.error.message}`)
+  }
+
+  const paymentRequestRow = await db.paymentRequest.findAll({
+    where: {
+      agreementNumber: paymentRequest.agreementNumber
+    }
+  })
+
+  if (paymentRequestRow && paymentRequestRow.length > 0) {
+    console.log('Duplicate payment request!')
+  } else {
+    const savedPaymentRequest = await db.paymentRequest.create(paymentRequest)
+    await processInvoiceLines(paymentRequest.invoiceLines, savedPaymentRequest.paymentRequestId)
+  }
+}
+
+async function processInvoiceLines (invoiceLines, paymentRequestId) {
+  if (invoiceLines.length > 0) {
+    for (const invoiceLine of invoiceLines) {
+      await db.invoiceLine.create({ paymentRequestId, ...invoiceLine })
+    }
+  }
+}
```

Comment:
This needs to run in a transaction or two services consuming different payment requests at the same time could result in a duplicate.

Including both the select and create in the same transaction will solve that problem.

You can also do `findOne` rather than `findAll` to allow you to simplify the `if`

This should work:

```
const transaction = await db.sequelize.transaction()
  try {
    const paymentRequestRow = await db.paymentRequest.findOne({
      transaction,
      where: {
        agreementNumber: paymentRequest.agreementNumber
      }
    })
    if (paymentRequestRow) {
    console.log('Duplicate payment request!')
  } else {
    const savedPaymentRequest = await db.paymentRequest.create(paymentRequest, {transaction} )
    await processInvoiceLines(paymentRequest.invoiceLines, savedPaymentRequest.paymentRequestId, {transaction})
  }
  } catch (error) {
    await transaction.rollback()
    throw (error)
  }
}

....

async function processInvoiceLines (invoiceLines, paymentRequestId, transaction) {
  if (invoiceLines.length > 0) {
    for (const invoiceLine of invoiceLines) {
      await db.invoiceLine.create({ paymentRequestId, ...invoiceLine },{transaction})
    }
  }
}
```